File: RouteGuard.cs
using System.Net.NetworkInformation;
using System.Diagnostics;
using System.Net;

namespace GatewayPolicy;

public enum MonitorState { Stopped, Running, Fixing }

public class RouteGuard
{
    private readonly AppConfig _config;
    private System.Threading.Timer? _timer;
    private int _mainIfIndex;
    private string _mainGateway = "";
    private volatile bool _fixing;
    private DateTime _lastFixTime = DateTime.MinValue;
    private string _routeTableCache = "";
    private DateTime _routeTableCacheTime = DateTime.MinValue;
    private string _lastNicLog = "";  // dedup: only log when NIC info changes
    private bool _wasHijacked = false;  // only log hijack on state transition
    private bool _steadyState = false;  // suppress logs when nothing changes

    // /2 counter-routes that override VPN's /1 hijack
    private static readonly string[] CounterRoutes = {
        "0.0.0.0/2", "64.0.0.0/2", "128.0.0.0/2", "192.0.0.0/2"
    };

    public MonitorState State { get; private set; } = MonitorState.Stopped;

    /// <summary>
    /// Current hijack status (true = VPN is hijacking routes).
    /// </summary>
    public bool IsHijacked => _wasHijacked;

    public event Action<string, LogLevel>? OnLog;
    public event Action<MonitorState>? OnStateChanged;
    public event Action? OnFixCompleted;
    public event Action<bool>? OnHijackChanged;

    public RouteGuard(AppConfig config)
    {
        _config = config;
    }

    public void Start()
    {
        if (State != MonitorState.Stopped) return;  // Already running or fixing

        CacheMainNic();

        State = MonitorState.Running;
        OnStateChanged?.Invoke(State);
        _timer = new System.Threading.Timer(_ => CheckLoop(), null,
            TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(_config.CheckIntervalSeconds));
        Log("Monitor started (v2.1: /2 counter-route strategy)", LogLevel.Info);
    }

    public void Stop()
    {
        _timer?.Dispose();
        _timer = null;
        _fixing = false;
        State = MonitorState.Stopped;
        OnStateChanged?.Invoke(State);
        Log("Monitor stopped", LogLevel.Info);
    }

    /// <summary>
    /// Remove all /2 counter-routes and restore original networking.
    /// Optionally stops monitoring first. Does NOT restart monitoring.
    /// </summary>
    public void RestoreNetwork()
    {
        // Stop monitoring to avoid re-adding routes during restoration
        if (State != MonitorState.Stopped)
        {
            _timer?.Dispose();
            _timer = null;
            _fixing = false;
            State = MonitorState.Stopped;
            OnStateChanged?.Invoke(State);
        }

        int removed = 0;

        // Remove /2 counter-routes
        foreach (var prefix in CounterRoutes)
        {
            RunRoute($"delete {prefix}");
            removed++;
        }

        // Also remove private routes to TAP
        foreach (var net in _config.PrivateNets)
        {
            RunRoute($"delete {net}");
            removed++;
        }

        // Remove custom routes
        foreach (var rule in _config.CustomRoutes)
        {
            RunRoute($"delete {rule.Key}");
            removed++;
        }

        _wasHijacked = false;
        _steadyState = false;

        Log($"Network restored — removed {removed} routes (counter-routes + private routes)", LogLevel.Info);
        OnHijackChanged?.Invoke(false);
        OnFixCompleted?.Invoke();
    }

    private bool CacheMainNic()
    {
        try
        {
            var nics = NetworkInterface.GetAllNetworkInterfaces()
                .Where(n => n.OperationalStatus == OperationalStatus.Up &&
                            n.NetworkInterfaceType != NetworkInterfaceType.Loopback &&
                            !IsVirtualAdapter(n))
                .ToList();

            foreach (var nic in nics)
            {
                var ipProps = nic.GetIPProperties();
                var gateway = ipProps.GatewayAddresses.FirstOrDefault(g => g.Address.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork);
                if (gateway != null && !string.IsNullOrEmpty(gateway.Address.ToString()))
                {
                    _mainIfIndex = nic.GetIPProperties().GetIPv4Properties().Index;
                    _mainGateway = gateway.Address.ToString();
                    // Only log when NIC info changes
                    var nicInfo = $"{nic.Name}|{_mainIfIndex}|{_mainGateway}";
                    if (_lastNicLog != nicInfo)
                    {
                        _lastNicLog = nicInfo;
                        Log($"Main NIC: {nic.Name} (if={_mainIfIndex}, gw={_mainGateway})", LogLevel.Info);
                    }
                    return true;
                }
            }
        }
        catch (Exception ex)
        {
            Log($"Detect main NIC error: {ex.Message}", LogLevel.Error);
        }
        return false;
    }

    private void CheckLoop()
    {
        if (_fixing) return;
        DoCheck();
    }

    public void DoCheck()
    {
        if (_fixing) return;

        try
        {
            CacheMainNic();

            // Primary check: def1 hijack routes (/1 routes override default route)
            bool hijacked = false;

            if (RouteExists("0.0.0.0/1") || RouteExists("128.0.0.0/1"))
            {
                hijacked = true;
                if (!_wasHijacked)
                {
                    Log("Detected VPN def1 hijack route (0.0.0.0/1 or 128.0.0.0/1)", LogLevel.Info);
                    _wasHijacked = true;
                    OnHijackChanged?.Invoke(true);
                }
            }
            else
            {
                if (_wasHijacked)
                {
                    Log("VPN hijack cleared — no longer detected", LogLevel.Info);
                    _wasHijacked = false;
                    _steadyState = false;
                    OnHijackChanged?.Invoke(false);
                }
            }

            // Secondary check: TAP has default route with very low metric
            foreach (var tap in GetManagedAdapters())
            {
                var metric = GetAdapterMetric(tap.IfIndex);
                if (metric > 0 && metric < 10)
                {
                    hijacked = true;
                    if (!_wasHijacked)
                    {
                        Log($"TAP adapter {tap.Name} has low metric default route (metric={metric})", LogLevel.Info);
                        _wasHijacked = true;
                    }
                }
            }

            if (hijacked) ApplyCounterRoutes();
        }
        catch (Exception ex)
        {
            Log($"Check error: {ex.Message}", LogLevel.Error);
        }
    }

    /// <summary>
    /// New strategy: Don't fight VPN. Let it push /1 routes.
    /// Instead, add /2 counter-routes pointing to main NIC.
    /// /2 is more specific than /1, so it wins without conflicting.
    /// VPN won't detect or fight back because it doesn't push /2 routes.
    /// </summary>
    private void ApplyCounterRoutes()
    {
        // Phase 1: Check if anything actually needs fixing (no state change)
        bool needFix = false;
        var routesToAdd = new List<(string cidr, string gateway, int ifIndex, int metric, string label)>();

        try
        {
            // Step 1: Check /2 counter-routes for public traffic → main NIC
            if (!string.IsNullOrEmpty(_mainGateway) && _mainIfIndex > 0)
            {
                foreach (var prefix in CounterRoutes)
                {
                    if (!CounterRouteExists(prefix, _mainIfIndex))
                    {
                        routesToAdd.Add((prefix, _mainGateway, _mainIfIndex, _config.MainMetric, "Counter-route"));
                        needFix = true;
                    }
                }
            }

            // Step 2: Check private networks → TAP
            var managedTaps = GetManagedAdapters();
            if (managedTaps.Count > 0)
            {
                int tapIf = managedTaps[0].IfIndex;
                string? tapGw = GetTapGateway(tapIf);

                if (!string.IsNullOrEmpty(tapGw))
                {
                    foreach (var net in _config.PrivateNets)
                    {
                        if (!PrivateRouteExists(net, tapIf))
                        {
                            routesToAdd.Add((net, tapGw, tapIf, 20, "Private route"));
                            needFix = true;
                        }
                    }
                }
            }

            // Step 3: Check custom route rules
            if (!string.IsNullOrEmpty(_mainGateway) && managedTaps.Count > 0)
            {
                int tapIf = managedTaps[0].IfIndex;
                string? tapGw = GetTapGateway(tapIf);

                foreach (var rule in _config.CustomRoutes)
                {
                    if (rule.Value == "tap" && !string.IsNullOrEmpty(tapGw))
                    {
                        if (!PrivateRouteExists(rule.Key, tapIf))
                        {
                            routesToAdd.Add((rule.Key, tapGw, tapIf, 20, "Custom route"));
                            needFix = true;
                        }
                    }
                    else if (rule.Value == "main")
                    {
                        if (!CounterRouteExists(rule.Key, _mainIfIndex))
                        {
                            routesToAdd.Add((rule.Key, _mainGateway, _mainIfIndex, _config.MainMetric, "Custom route"));
                            needFix = true;
                        }
                    }
                }
            }

            // Phase 2: Only enter Fixing state if we actually need to add routes
            if (!needFix)
            {
                if (!_steadyState)
                {
                    Log("All routes verified OK — monitoring (steady state)", LogLevel.Info);
                    _steadyState = true;
                }
                return;  // Stay in Running state, no flicker
            }

            _fixing = true;
            State = MonitorState.Fixing;
            OnStateChanged?.Invoke(State);
            _steadyState = false;

            // Phase 3: Actually add the missing routes
            foreach (var r in routesToAdd)
            {
                RunRoute($"add {r.cidr} {r.gateway} if {r.ifIndex} metric {r.metric}");
                Log($"{r.label} ADD: {r.cidr} -> if={r.ifIndex} gw={r.gateway} metric={r.metric}", LogLevel.Info);
            }

            _config.TotalFixes++;
            _config.LastFixTime = DateTime.Now;
            _config.Save();
            Log($"Counter-routes applied (total fixes: {_config.TotalFixes})", LogLevel.Info);
            _lastFixTime = DateTime.Now;

            OnFixCompleted?.Invoke();

            if (_config.BarkEnabled && !string.IsNullOrEmpty(_config.BarkServer))
            {
                _ = SendBarkNotification();
            }
        }
        catch (Exception ex)
        {
            Log($"Counter-route error: {ex.Message}", LogLevel.Error);
        }
        finally
        {
            if (_fixing)
            {
                _fixing = false;
                // Only restore to Running if not stopped by user during fix
                if (State == MonitorState.Fixing)
                {
                    State = MonitorState.Running;
                    OnStateChanged?.Invoke(State);
                }
            }
        }
    }

    // --- Helper methods ---

    private static bool IsVirtualAdapter(NetworkInterface nic)
    {
        var desc = nic.Description;
        string[] keywords = { "TAP", "VPN", "WireGuard", "Tunnel", "Virtual", "Hyper-V", "EricVPN" };
        return keywords.Any(k => desc.Contains(k, StringComparison.OrdinalIgnoreCase));
    }

    private List<TapAdapter> GetManagedAdapters()
    {
        string[] managedKeywords = { "TAP", "VPN", "Virtual", "Hyper-V", "EricVPN" };
        string[] ignoreKeywords = { "WireGuard" };

        return NetworkInterface.GetAllNetworkInterfaces()
            .Where(n => n.OperationalStatus == OperationalStatus.Up)
            .Where(n => managedKeywords.Any(k => n.Description.Contains(k, StringComparison.OrdinalIgnoreCase)))
            .Where(n => !ignoreKeywords.Any(k => n.Description.Contains(k, StringComparison.OrdinalIgnoreCase)))
            .Select(n => new TapAdapter
            {
                Name = n.Name,
                IfIndex = n.GetIPProperties().GetIPv4Properties().Index,
                Description = n.Description
            })
            .ToList();
    }

    // --- Route table parsing helpers ---

    /// <summary>
    /// Get cached route table output (cache for 2 seconds to avoid spamming route.exe)
    /// </summary>
    private string GetRouteTable()
    {
        if ((DateTime.Now - _routeTableCacheTime).TotalSeconds < 2 && !string.IsNullOrEmpty(_routeTableCache))
            return _routeTableCache;
        _routeTableCache = RunRoute("print -4");
        _routeTableCacheTime = DateTime.Now;
        return _routeTableCache;
    }

    /// <summary>
    /// Map ifIndex to the interface's IPv4 address (used for matching route print output,
    /// since route print shows interface IP, not ifIndex).
    /// </summary>
    private string? GetInterfaceIp(int ifIndex)
    {
        try
        {
            var nic = NetworkInterface.GetAllNetworkInterfaces()
                .FirstOrDefault(n => n.GetIPProperties().GetIPv4Properties().Index == ifIndex);
            if (nic == null) return null;
            var ip = nic.GetIPProperties().UnicastAddresses
                .FirstOrDefault(a => a.Address.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork);
            return ip?.Address.ToString();
        }
        catch { return null; }
    }

    /// <summary>
    /// Convert CIDR prefix (e.g. "0.0.0.0/1") to Windows route print format:
    /// ("0.0.0.0", "128.0.0.0") — Network Destination and Netmask.
    /// </summary>
    private static (string dest, string mask) CidrToRoutePrint(string cidr)
    {
        var parts = cidr.Split('/');
        if (parts.Length != 2 || !int.TryParse(parts[1], out int prefixLen) || prefixLen < 0 || prefixLen > 32)
            return (parts[0], "0.0.0.0"); // fallback

        // Build mask in network byte order (big-endian) to avoid IPAddress(uint) endianness issues
        uint maskVal = prefixLen == 0 ? 0 : 0xFFFFFFFFu << (32 - prefixLen);
        byte[] maskBytes = new byte[4];
        maskBytes[0] = (byte)((maskVal >> 24) & 0xFF);
        maskBytes[1] = (byte)((maskVal >> 16) & 0xFF);
        maskBytes[2] = (byte)((maskVal >> 8) & 0xFF);
        maskBytes[3] = (byte)(maskVal & 0xFF);
        var maskAddr = new IPAddress(maskBytes);
        // Network destination = IP AND mask (both in network byte order)
        var ipBytes = IPAddress.Parse(parts[0]).GetAddressBytes();
        var destBytes = new byte[4];
        for (int i = 0; i < 4; i++) destBytes[i] = (byte)(ipBytes[i] & maskBytes[i]);
        return (new IPAddress(destBytes).ToString(), maskAddr.ToString());
    }

    /// <summary>
    /// Check if a CIDR route exists in route table (any interface).
    /// Matches against Windows route print format: Dest  Mask  Gateway  InterfaceIP  Metric
    /// </summary>
    private bool RouteExists(string cidr)
    {
        try
        {
            var output = GetRouteTable();
            var (dest, mask) = CidrToRoutePrint(cidr);
            var lines = output.Split('\n');
            return lines.Any(l =>
            {
                var parts = l.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                return parts.Length >= 5 && parts[0] == dest && parts[1] == mask;
            });
        }
        catch { return false; }
    }

    /// <summary>
    /// Check if a counter-route (CIDR → specific ifIndex) exists in route table.
    /// Uses interface IP mapping since route print shows IP not ifIndex.
    /// </summary>
    private bool CounterRouteExists(string cidr, int ifIndex)
    {
        return RouteExistsOnInterface(cidr, ifIndex);
    }

    /// <summary>
    /// Check if a private network route (CIDR → TAP ifIndex) exists in route table.
    /// </summary>
    private bool PrivateRouteExists(string cidr, int ifIndex)
    {
        return RouteExistsOnInterface(cidr, ifIndex);
    }

    /// <summary>
    /// Check if a CIDR route exists on a specific interface (by ifIndex).
    /// Resolves ifIndex to interface IP for route print matching.
    /// </summary>
    private bool RouteExistsOnInterface(string cidr, int ifIndex)
    {
        try
        {
            var output = GetRouteTable();
            var (dest, mask) = CidrToRoutePrint(cidr);
            var ifIp = GetInterfaceIp(ifIndex);
            var lines = output.Split('\n');
            return lines.Any(l =>
            {
                var parts = l.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length < 5 || parts[0] != dest || parts[1] != mask)
                    return false;
                // Match by interface IP if available, otherwise match by gateway
                if (!string.IsNullOrEmpty(ifIp))
                    return parts[3] == ifIp;
                return true; // If we can't resolve IP, just match dest+mask
            });
        }
        catch { return false; }
    }

    private string? GetTapGateway(int ifIndex)
    {
        try
        {
            var ifIp = GetInterfaceIp(ifIndex);
            var output = GetRouteTable();
            var lines = output.Split('\n');
            foreach (var line in lines)
            {
                var parts = line.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length < 5) continue;
                // Match lines on this interface (by interface IP column)
                if (!string.IsNullOrEmpty(ifIp) && parts[3] != ifIp) continue;
                var gw = parts[2];
                if (gw != "On-link" && System.Net.IPAddress.TryParse(gw, out var ip) && !ip.Equals(System.Net.IPAddress.Any))
                    return gw;
            }
        }
        catch { }
        return null;
    }

    private int GetAdapterMetric(int ifIndex)
    {
        try
        {
            var ifIp = GetInterfaceIp(ifIndex);
            var output = GetRouteTable();
            var lines = output.Split('\n');
            foreach (var line in lines)
            {
                var parts = line.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length < 5) continue;
                // Look for default route (0.0.0.0/0) on this interface
                if (parts[0] == "0.0.0.0" && parts[1] == "0.0.0.0")
                {
                    if (!string.IsNullOrEmpty(ifIp) && parts[3] != ifIp) continue;
                    if (int.TryParse(parts[4], out var m))
                        return m;
                }
            }
        }
        catch { }
        return -1;
    }

    private string RunRoute(string args)
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "route.exe",
                Arguments = args,
                CreateNoWindow = true,
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            };
            using var proc = Process.Start(psi);
            if (proc == null) return "";
            var output = proc.StandardOutput.ReadToEnd();
            proc.WaitForExit(5000);
            return output;
        }
        catch { return ""; }
    }

    private async Task SendBarkNotification()
    {
        try
        {
            var server = _config.BarkServer.TrimEnd('/');
            var key = _config.BarkDeviceKey;
            var title = Uri.EscapeDataString(_config.BarkTitle);
            var body = Uri.EscapeDataString($"Route fix #{_config.TotalFixes} at {DateTime.Now:HH:mm:ss}");
            var url = $"{server}/{key}/{title}/{body}";
            if (!string.IsNullOrEmpty(_config.BarkSound))
                url += $"?sound={_config.BarkSound}";

            using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(5) };
            await http.GetAsync(url);
            Log("Bark notification sent", LogLevel.Info);
        }
        catch (Exception ex)
        {
            Log($"Bark send error: {ex.Message}", LogLevel.Warning);
        }
    }

    private void Log(string message, LogLevel level)
    {
        OnLog?.Invoke(message, level);
    }
}

public class TapAdapter
{
    public string Name { get; set; } = "";
    public int IfIndex { get; set; }
    public string Description { get; set; } = "";
}

public enum LogLevel { Info, Warning, Error }

File: src/route_guard.rs
use crate::config::AppConfig;
use crate::logger::Logger;
use crate::network;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

#[derive(Clone, Copy, PartialEq, Debug)]
pub enum MonitorState {
    Stopped,
    Running,
    Fixing,
}

const COUNTER_ROUTES: &[&str] = &["0.0.0.0/2", "64.0.0.0/2", "128.0.0.0/2", "192.0.0.0/2"];

struct GuardInner {
    main_if_index: u32,
    main_if_ip: String,
    main_gateway: String,
    main_nic_name: String,
    fixing: bool,
    was_hijacked: bool,
    steady_state: bool,
    last_nic_log: String,
    state: MonitorState,
    running: bool,
}

pub struct RouteGuard {
    inner: Arc<Mutex<GuardInner>>,
    config: Arc<Mutex<AppConfig>>,
    logger: Arc<Logger>,
}

#[derive(Clone)]
pub struct GuardStatus {
    pub state: MonitorState,
    pub is_hijacked: bool,
    pub main_nic: String,
    pub tap_nic: String,
    pub total_fixes: u64,
    pub last_fix: String,
}

impl RouteGuard {
    pub fn new(config: Arc<Mutex<AppConfig>>, logger: Arc<Logger>) -> Self {
        RouteGuard {
            inner: Arc::new(Mutex::new(GuardInner {
                main_if_index: 0,
                main_if_ip: String::new(),
                main_gateway: String::new(),
                main_nic_name: String::new(),
                fixing: false,
                was_hijacked: false,
                steady_state: false,
                last_nic_log: String::new(),
                state: MonitorState::Stopped,
                running: false,
            })),
            config,
            logger,
        }
    }

    pub fn get_status(&self) -> GuardStatus {
        let inner = self.inner.lock().unwrap();
        let config = self.config.lock().unwrap();
        GuardStatus {
            state: inner.state,
            is_hijacked: inner.was_hijacked,
            main_nic: inner.main_nic_name.clone(),
            tap_nic: {
                let adapters = network::get_adapters();
                let taps = network::get_managed_adapters(&adapters);
                taps.first().map(|t| t.name.clone()).unwrap_or_default()
            },
            total_fixes: config.total_fixes,
            last_fix: config.last_fix_time.clone().unwrap_or_default(),
        }
    }

    pub fn start(&self) {
        let mut inner = self.inner.lock().unwrap();
        if inner.running {
            return;
        }
        inner.running = true;
        inner.state = MonitorState::Running;
        drop(inner);

        self.cache_main_nic();
        self.logger.info("Monitor started (v4.0.0: /2 counter-route strategy, Rust)");

        // Spawn monitoring thread
        let inner = self.inner.clone();
        let config = self.config.clone();
        let logger = self.logger.clone();

        thread::spawn(move || {
            thread::sleep(Duration::from_secs(2)); // initial delay
            loop {
                let should_run = {
                    let g = inner.lock().unwrap();
                    g.running
                };
                if !should_run { break; }

                let check_interval = config.lock().unwrap().check_interval_seconds;
                do_check(&inner, &config, &logger);
                thread::sleep(Duration::from_secs(check_interval.max(1) as u64));
            }
        });
    }

    pub fn stop(&self) {
        let mut inner = self.inner.lock().unwrap();
        inner.running = false;
        inner.state = MonitorState::Stopped;
        drop(inner);
        self.logger.info("Monitor stopped");
    }

    pub fn do_check_now(&self) {
        do_check(&self.inner, &self.config, &self.logger);
    }

    pub fn restore_network(&self) {
        self.stop();

        let mut removed = 0;
        for prefix in COUNTER_ROUTES {
            network::run_route(&format!("delete {}", prefix));
            removed += 1;
        }

        let config = self.config.lock().unwrap();
        for net in &config.private_nets {
            network::run_route(&format!("delete {}", net));
            removed += 1;
        }
        for key in config.custom_routes.keys() {
            network::run_route(&format!("delete {}", key));
            removed += 1;
        }
        drop(config);

        let mut inner = self.inner.lock().unwrap();
        inner.was_hijacked = false;
        inner.steady_state = false;
        drop(inner);

        self.logger.info(&format!("Network restored — removed {} routes", removed));
    }

    fn cache_main_nic(&self) {
        let adapters = network::get_adapters();
        if let Some(nic) = network::get_main_nic(&adapters) {
            let nic_info = format!("{}|{}|{}", nic.name, nic.if_index, nic.gateway);
            let mut inner = self.inner.lock().unwrap();
            inner.main_if_index = nic.if_index;
            inner.main_if_ip = nic.ip.clone();
            inner.main_gateway = nic.gateway.clone();
            inner.main_nic_name = nic.name.clone();

            if inner.last_nic_log != nic_info {
                inner.last_nic_log = nic_info;
                drop(inner);
                self.logger.info(&format!("Main NIC: {} (if={}, gw={})", nic.name, nic.if_index, nic.gateway));
            }
        } else {
            self.logger.error("Could not detect main NIC");
        }
    }
}

fn do_check(inner: &Arc<Mutex<GuardInner>>, config: &Arc<Mutex<AppConfig>>, logger: &Arc<Logger>) {
    {
        let g = inner.lock().unwrap();
        if g.fixing || !g.running {
            return;
        }
    }

    // Refresh main NIC info
    let adapters = network::get_adapters();
    if let Some(nic) = network::get_main_nic(&adapters) {
        let nic_info = format!("{}|{}|{}", nic.name, nic.if_index, nic.gateway);
        let mut g = inner.lock().unwrap();
        g.main_if_index = nic.if_index;
        g.main_if_ip = nic.ip.clone();
        g.main_gateway = nic.gateway.clone();
        g.main_nic_name = nic.name.clone();
        let should_log = g.last_nic_log != nic_info;
        if should_log {
            g.last_nic_log = nic_info.clone();
        }
        drop(g);
        if should_log {
            logger.info(&format!("Main NIC: {} (if={}, gw={})", nic.name, nic.if_index, nic.gateway));
        }
    }

    // Check for def1 hijack
    let mut hijacked = false;
    if network::route_exists("0.0.0.0/1") || network::route_exists("128.0.0.0/1") {
        hijacked = true;
        let mut g = inner.lock().unwrap();
        if !g.was_hijacked {
            g.was_hijacked = true;
            drop(g);
            logger.info("Detected VPN def1 hijack route (0.0.0.0/1 or 128.0.0.0/1)");
        }
    } else {
        let mut g = inner.lock().unwrap();
        if g.was_hijacked {
            g.was_hijacked = false;
            g.steady_state = false;
            drop(g);
            logger.info("VPN hijack cleared — no longer detected");
        }
    }

    // Secondary check: TAP has low metric default route
    let taps = network::get_managed_adapters(&adapters);
    for tap in &taps {
        let metric = network::get_adapter_metric(&tap.ip);
        if metric > 0 && metric < 10 {
            hijacked = true;
            let mut g = inner.lock().unwrap();
            if !g.was_hijacked {
                g.was_hijacked = true;
                drop(g);
                logger.info(&format!("TAP adapter {} has low metric default route (metric={})", tap.name, metric));
            }
        }
    }

    if hijacked {
        apply_counter_routes(inner, config, logger);
    }
}

fn apply_counter_routes(inner: &Arc<Mutex<GuardInner>>, config: &Arc<Mutex<AppConfig>>, logger: &Arc<Logger>) {
    let g = inner.lock().unwrap();
    if g.fixing { return; }
    let main_gw = g.main_gateway.clone();
    let main_if = g.main_if_index;
    let main_if_ip = g.main_if_ip.clone();
    drop(g);

    let adapters = network::get_adapters();
    let taps = network::get_managed_adapters(&adapters);

    // Phase 1: Check what needs fixing
    let mut need_fix = false;
    let mut routes_to_add: Vec<(String, String, u32, u32, &str)> = Vec::new();

    let cfg = config.lock().unwrap();
    let main_metric = cfg.main_metric;
    let private_nets = cfg.private_nets.clone();
    let custom_routes = cfg.custom_routes.clone();
    drop(cfg);

    // Step 1: Check /2 counter-routes
    if !main_gw.is_empty() && main_if > 0 {
        for prefix in COUNTER_ROUTES {
            if !network::route_exists_on_interface(prefix, &main_if_ip) {
                routes_to_add.push((prefix.to_string(), main_gw.clone(), main_if, main_metric, "Counter-route"));
                need_fix = true;
            }
        }
    }

    // Step 2: Check private networks -> TAP
    if !taps.is_empty() {
        let tap = taps[0];
        let tap_gw = network::get_tap_gateway(&tap.ip);
        if let Some(ref gw) = tap_gw {
            for net in &private_nets {
                if !network::route_exists_on_interface(net, &tap.ip) {
                    routes_to_add.push((net.clone(), gw.clone(), tap.if_index, 20, "Private route"));
                    need_fix = true;
                }
            }
        }
    }

    // Step 3: Check custom routes
    if !main_gw.is_empty() && !taps.is_empty() {
        let tap = taps[0];
        let tap_gw = network::get_tap_gateway(&tap.ip);

        for (prefix, via) in &custom_routes {
            if via == "tap" {
                if let Some(ref gw) = tap_gw {
                    if !network::route_exists_on_interface(prefix, &tap.ip) {
                        routes_to_add.push((prefix.clone(), gw.clone(), tap.if_index, 20, "Custom route"));
                        need_fix = true;
                    }
                }
            } else if via == "main" {
                if !network::route_exists_on_interface(prefix, &main_if_ip) {
                    routes_to_add.push((prefix.clone(), main_gw.clone(), main_if, main_metric, "Custom route"));
                    need_fix = true;
                }
            }
        }
    }

    if !need_fix {
        let mut g = inner.lock().unwrap();
        if !g.steady_state {
            g.steady_state = true;
            drop(g);
            logger.info("All routes verified OK — monitoring (steady state)");
        }
        return;
    }

    // Phase 2: Enter Fixing state
    {
        let mut g = inner.lock().unwrap();
        g.fixing = true;
        g.state = MonitorState::Fixing;
        g.steady_state = false;
    }

    // Phase 3: Actually add routes
    for (cidr, gw, ifidx, metric, label) in &routes_to_add {
        network::run_route(&format!("add {} {} if {} metric {}", cidr, gw, ifidx, metric));
        logger.info(&format!("{} ADD: {} -> if={} gw={} metric={}", label, cidr, ifidx, gw, metric));
    }

    // Update config
    {
        let mut cfg = config.lock().unwrap();
        cfg.total_fixes += 1;
        cfg.last_fix_time = Some(chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string());
        cfg.save();
        let total = cfg.total_fixes;
        let bark = cfg.bark_enabled;
        let bark_server = cfg.bark_server.clone();
        let bark_key = cfg.bark_device_key.clone();
        let bark_title = cfg.bark_title.clone();
        drop(cfg);

        logger.info(&format!("Counter-routes applied (total fixes: {})", total));

        if bark && !bark_server.is_empty() && !bark_key.is_empty() {
            send_bark(&bark_server, &bark_key, &bark_title, total, logger);
        }
    }

    // Restore to Running
    {
        let mut g = inner.lock().unwrap();
        g.fixing = false;
        if g.state == MonitorState::Fixing {
            g.state = MonitorState::Running;
        }
    }
}

fn send_bark(server: &str, key: &str, title: &str, total: u64, logger: &Arc<Logger>) {
    let server = server.trim_end_matches('/');
    let now = chrono::Local::now().format("%H:%M:%S");
    let title_enc = url_encode(title);
    let body_enc = url_encode(&format!("Route fix #{} at {}", total, now));
    let url = format!("{}/{}/{}/{}", server, key, title_enc, body_enc);

    match reqwest::blocking::get(&url) {
        Ok(r) => {
            if r.status().is_success() {
                logger.info("Bark notification sent");
            } else {
                logger.warn(&format!("Bark response: {}", r.status()));
            }
        }
        Err(e) => logger.warn(&format!("Bark send error: {}", e)),
    }
}

fn url_encode(s: &str) -> String {
    let mut result = String::new();
    for byte in s.bytes() {
        if byte.is_ascii_alphanumeric() || b"-_.~".contains(&byte) {
            result.push(byte as char);
        } else {
            result.push_str(&format!("%{:02X}", byte));
        }
    }
    result
}
